addons/mail/models/mail_thread.py

```diff
@@ -430,8 +430,10 @@ def _prepare_tracking(self, fields):
         fnames = self._get_tracked_fields().intersection(fields)
         if not fnames:
             return
-        func = self.browse()._finalize_tracking
-        [initial_values] = self.env.cr.precommit.add(func, dict)
+        mail_tracking = self.env.cr.precommit.data.setdefault('mail_tracking', {})
+        initial_values = mail_tracking.setdefault(self._name, {})
+        if not initial_values:
+            self.env.cr.precommit.add(self._finalize_tracking)
```

Comment:
This really seems like a verbose and error-prone API to use. Especially with the idea of checking if the data dict is falsy then adding a callback, as that seems easy to get wrong (e.g. if you just need a callback with no additional data).

Maybe have `hook.add(key, hookfunc) -> data`?

`data` could be either `hook.data` or `hook.data.setdefault(key, {})` though the latter case is significantly less flexible (no way to make your "parametrisation" a list) it's also safer. Not sure which is better. I guess the shortcut could just go with "parameters are dicts" and you can either not use the shortcut if you need something more complicated, or set your list (or set or whatever) as some arbitrary key inside your data dict. Or the shortcut used as the initial value if the key's unset I guess.

With that, the code would become
```suggestion
        initial_values = self.env.cr.precommit.add('mail_tracking', self._finalize_tracking).setdefault(self._name, {})
```
And we'd rename `add` to e.g. `add_callback` or `add_func` as we'd expect it to be used quite a bit less. But `add` would be something relatively straightforward e.g.
```python
def add(self, key, func, default=None):
    if key in self.data:
        return self.data[key]

    self.add(func)
    return self.data.setdefault(key, {} if default is None else default)
```
(maybe make `default` a placeholder object so we can we the data holder to `None`, or maybe make `default` a callable as in defaultdict, that's probably the most troublesome part of the interface)